From 55e3567cf3b3994b74af4e17bff01da04ab99e6b Mon Sep 17 00:00:00 2001
From: Zhou Wang <wangzhou1@hisilicon.com>
Date: Mon, 20 Oct 2025 21:30:02 +0800
Subject: [PATCH 777/960] downstream: KVM: hisi_virt: tlbi: Fix wrong CPU aff3
 conversion between MPIDR and SYS_LSUDVMBM_EL2

driver inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/ICURZK

---------------------------------------------------------------------

TLBI broadcast CPU bitmap should be set in SYS_LSUDVMBM_EL2. Now we make a
mistake when doing the conversion between MPIDR and SYS_LSUDVMBM_EL2.

Fields of Die ID and Socket ID in Aff3 are different between MPIDR and
SYS_LSUDVMBM_EL2 in HIP12, however, they are same in current wrong logic.

This patch fixes this problem.

Fixes: 2671ba221968 ("kvm: hisi_virt: Update TLBI broadcast feature for hip12")
Signed-off-by: Zhou Wang <wangzhou1@hisilicon.com>
Signed-off-by: Jian Cai <caijian11@h-partners.com>
Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 arch/arm64/kvm/hisilicon/hisi_virt.c | 25 ++++++++++++++++++++++---
 arch/arm64/kvm/hisilicon/hisi_virt.h | 25 ++++++++++++++++++++-----
 2 files changed, 42 insertions(+), 8 deletions(-)

diff --git a/arch/arm64/kvm/hisilicon/hisi_virt.c b/arch/arm64/kvm/hisilicon/hisi_virt.c
index 48b644e4b3f3..cf22f247f00f 100644
--- a/arch/arm64/kvm/hisilicon/hisi_virt.c
+++ b/arch/arm64/kvm/hisilicon/hisi_virt.c
@@ -8,6 +8,7 @@
 #include <linux/init.h>
 #include <linux/kvm_host.h>
 #include "hisi_virt.h"
+#include <linux/bitfield.h>
 
 static enum hisi_cpu_type cpu_type = UNKNOWN_HI_TYPE;
 
@@ -475,12 +476,27 @@ static void kvm_update_vm_lsudvmbm(struct kvm *kvm)
 	kvm->arch.tlbi_dvmbm = val;
 }
 
+static u64 convert_aff3_to_die_hip12(u64 aff3)
+{
+	/*
+	 * On HIP12, we use 4 bits to represent a die in SYS_LSUDVMBM_EL2.
+	 *
+	 * die1: socket ID (bits[60:59]) + die ID (bits[58:57])
+	 * die2: socket ID (bits[56:55]) + die ID (bits[54:53])
+	 *
+	 * We therefore need to properly encode Aff3 into it.
+	 */
+	return FIELD_GET(MPIDR_AFF3_SOCKET_ID_MASK, aff3) << 2 |
+	       FIELD_GET(MPIDR_AFF3_DIE_ID_MASK, aff3);
+}
+
 static void kvm_update_vm_lsudvmbm_hip12(struct kvm *kvm)
 {
 	u64 mpidr, aff3, aff2;
 	u64 vm_aff3s[DVMBM_MAX_DIES_HIP12];
 	u64 val;
 	int cpu, nr_dies;
+	u64 die1, die2;
 
 	nr_dies = kvm_dvmbm_get_dies_info(kvm, vm_aff3s, DVMBM_MAX_DIES_HIP12);
 	if (nr_dies > 2) {
@@ -489,8 +505,9 @@ static void kvm_update_vm_lsudvmbm_hip12(struct kvm *kvm)
 	}
 
 	if (nr_dies == 1) {
+		die1 = convert_aff3_to_die_hip12(vm_aff3s[0]);
 		val = DVMBM_RANGE_ONE_DIE << DVMBM_RANGE_SHIFT	|
-		      vm_aff3s[0] << DVMBM_DIE1_VDIE_SHIFT_HIP12;
+		      die1 << DVMBM_DIE1_SHIFT_HIP12;
 
 		/* fulfill bits [11:6] */
 		for_each_cpu(cpu, kvm->arch.sched_cpus) {
@@ -504,10 +521,12 @@ static void kvm_update_vm_lsudvmbm_hip12(struct kvm *kvm)
 	}
 
 	/* nr_dies == 2 */
+	die1 = convert_aff3_to_die_hip12(vm_aff3s[0]);
+	die2 = convert_aff3_to_die_hip12(vm_aff3s[1]);
 	val = DVMBM_RANGE_TWO_DIES << DVMBM_RANGE_SHIFT	|
 	      DVMBM_GRAN_CLUSTER << DVMBM_GRAN_SHIFT	|
-	      vm_aff3s[0] << DVMBM_DIE1_VDIE_SHIFT_HIP12    |
-	      vm_aff3s[1] << DVMBM_DIE2_VDIE_SHIFT_HIP12;
+	      die1 << DVMBM_DIE1_SHIFT_HIP12		|
+	      die2 << DVMBM_DIE2_SHIFT_HIP12;
 
 	/* and fulfill bits [11:0] */
 	for_each_cpu(cpu, kvm->arch.sched_cpus) {
diff --git a/arch/arm64/kvm/hisilicon/hisi_virt.h b/arch/arm64/kvm/hisilicon/hisi_virt.h
index 2609e360a254..36d77f89c5fe 100644
--- a/arch/arm64/kvm/hisilicon/hisi_virt.h
+++ b/arch/arm64/kvm/hisilicon/hisi_virt.h
@@ -43,7 +43,7 @@ enum hisi_cpu_type {
 #define TOTEM_B_ID 3
 
 /*
- * MPIDR_EL1 layout on HIP10
+ * MPIDR_EL1 layout on HIP10/HIP10C
  *
  * Aff3[7:3]	- socket ID	[0-15]
  * Aff3[2:0]	- die ID	[1,3]
@@ -69,12 +69,27 @@ enum hisi_cpu_type {
 
 #define DVMBM_MAX_DIES			32
 
-/* HIP12 */
-#define DVMBM_DIE1_VDIE_SHIFT_HIP12		57
-#define DVMBM_DIE2_VDIE_SHIFT_HIP12		53
+/*
+ * MPIDR_EL1 layout on HIP12
+ *
+ * Aff3[4:3]	- socket ID	[0-3]
+ * Aff3[2:0]	- die ID	[0,1]
+ * Aff2[2:0]	- cluster ID	[0-5]
+ * Aff1[3:0]	- core ID	[0-15]
+ * Aff0[0]	- thread ID	[0,1]
+ *
+ * On HIP12, cpu die is named as vdie. Actually,
+ * vdie is equivalent to cpu die. Here use die
+ * to describe vdie.
+ */
+
+#define MPIDR_AFF3_SOCKET_ID_MASK		GENMASK(4, 3)
+#define MPIDR_AFF3_DIE_ID_MASK			GENMASK(2, 0)
+#define DVMBM_DIE1_SHIFT_HIP12			57
+#define DVMBM_DIE2_SHIFT_HIP12			53
 #define DVMBM_DIE1_CLUSTER_SHIFT_HIP12		6
 #define DVMBM_DIE2_CLUSTER_SHIFT_HIP12		0
-#define DVMBM_MAX_DIES_HIP12				8
+#define DVMBM_MAX_DIES_HIP12			8
 
 void probe_hisi_cpu_type(void);
 bool hisi_ncsnp_supported(void);
-- 
2.43.0

